src/kudu/consensus/consensus_queue.cc

```diff
@@ -1051,6 +1062,51 @@ void PeerMessageQueue::PromoteIfNeeded(TrackedPeer* peer, const TrackedPeer& pre
   }
 }
 
+bool PeerMessageQueue::BasicChecksOKToTransferUnlocked(
+    const TrackedPeer& peer, RaftPeerPB **peer_pb_ptr) {
+  DCHECK(queue_lock_.is_locked());
+
+  // This check is redundant for ResponseFromPeer common path
+  if (PREDICT_FALSE(queue_state_.state != kQueueOpen)) {
+    return false;
+  }
+
+  if (queue_state_.mode != PeerMessageQueue::LEADER ||
+      peer.last_exchange_status != PeerStatus::OK) {
+    return false;
+  }
+
+  Status s = GetRaftConfigMember(DCHECK_NOTNULL(queue_state_.active_config.get()),
+                                 peer.uuid(), peer_pb_ptr);
+  if (!s.ok() || (*peer_pb_ptr)->member_type() != RaftPeerPB::VOTER) {
+    return false;
+  }
+
+  return true;
+}
+
+bool PeerMessageQueue::PeerTransferLeadershipImmediatelyUnlocked(
+    const std::string& peer_uuid) {
+  DCHECK(queue_lock_.is_locked());
+  TrackedPeer* peer = FindPtrOrNull(peers_map_, peer_uuid);
+  if (PREDICT_FALSE(peer == nullptr)) {
+    return false;
+  }
+
+  RaftPeerPB *peer_pb = nullptr;
+  if (!BasicChecksOKToTransferUnlocked(*peer, &peer_pb)) {
+    return false;
+  }
+
+  bool peer_caught_up =
+      !OpIdEquals(peer->last_received, MinimumOpId()) &&
+      OpIdEquals(peer->last_received, queue_state_.last_appended);
```

Comment:
Small comment about why only this condition is not sufficient?

Reply:
I think the MinimumOpId change maybe redundant, if thats what you mean but I am being risk averse as I have copied the check.